sys/fork.c

```diff
@@ -0,0 +1,64 @@
+#include <fork.h>
+#include <thread.h>
+#include <sync.h>
+#include <filedesc.h>
+#include <sched.h>
+#include <stdc.h>
+
+int do_fork() {
+  thread_t *td = thread_self();
+
+  /* Cannot fork non-user threads. */
+  assert(td->td_uspace);
+
+  thread_t *newtd = thread_create(td->td_name, NULL, NULL);
+
+  /* Prevent preemption, so that the copied thread won't run while we're copying
+     it. */
+  critical_enter();
```

Comment:
You perform several steps with interrupts turned off. For now they seem to be rather cheap, however I cannot conceive of copying address space under critical section. This has to be addressed.